Refactoring

yuripourre opened the pull request.

yuripourre commented:
@gotev Build passing, thanks for updating travis.

gotev commented:
@yuripourre thanks for the PR! 🎉 

Can you please explain the reasons of this refactoring and improvements it gives?

yuripourre commented:
@gotev thanks for commenting, this PR only does small refactoring, my final goal was to completely separate UI and Log from the main logic since I don't need it for my projects. The next changes I am planning would break compatibility with the current structure then I decided to do in two steps.

I also opened a PR to know if the project is actively maintained. 😅

gotev commented:
@yuripourre version 1.x is not actively maintained. During lockdown I started a complete redesign and rewrite in Kotlin for [2.0](https://github.com/gotev/android-speech/tree/2.0), using also coroutines. The whole rewrite is just at its start, then I haven't had more time to work on it. There's a first working implementation of the TTS engine. Speech-To-Text still needs to be implemented. In my opinion, 1.x is not worth refactoring. It just takes too much time and there are flawed design decisions which needs to be dropped. The plan is to make a coroutine wrapper of the system capabilities which is easy to use, modular and composable. You're welcome to be part of it.

yuripourre commented:
@gotev I see, thanks for the explanation and the invite. Seems promising but I am afraid I will not have enough time to help since it's kotlin, at least at the moment.

 I will be working on the fork I made instead.

Feel free to close the PR.

gotev commented:
@yuripourre I don't have much time either at the moment unfortunately, but I'm curious about your refactoring 😃. Who knows, 2.x may not be what I originally imagined, but that's the good part of this community! Ideas and different point of views are always welcome 😉. Feel free to continue your work in Java in your fork and don't be afraid to introduce non retro-compatible stuff. I'm available for design feedbacks, advice and code review.

yuripourre commented:
@gotev what do you think about the current approach?
The way it is now, I can create my own `SpeechRecognitionListener` without logs nor UI

The next step will be extracting a listener to handle TTS

gotev commented on speech/src/main/java/net/gotev/speech/Speech.java:
```diff
@@ -28,236 +22,39 @@
-            mTextToSpeech.setPitch(mTtsPitch);
-            mTextToSpeech.setSpeechRate(mTtsRate);
-        }
+    public Speech(final Context context, final String callingPackage, TextToSpeech.OnInitListener onInitListener, SpeechRecognitionListener speechRecognitionListener) {
```
`speechRecognitionListener` is not used

yuripourre replied:
Oops, copy+paste bug.

gotev commented:
@yuripourre good job! Only thing I noticed by looking at the diff is this: https://github.com/gotev/android-speech/pull/42/files#r506967601

This goes in the same direction as I planned for 2.0, where you could use TTS only, Speech-To-Text only or both combined and orchestrated, without initializing unneeded stuff. Also, with your design you've now allowed subclassing. This opens for more customizations in how the TTS and Speech-To-Text are used.

yuripourre commented:
@gotev Thank you for the feedback, just fixed your comment. Yes, I tried to make the library as flexible as possible (and now proguard can probably reduce the app size a bit as well).

Some names could be changed, I know, but it can be done later.

Do you think this PR should be merged?

gotev commented:
@yuripourre given a more in depth look at the PR from a `can it be merged as it is?` perspective. As such you will notice I'm trying to reduce or eliminate the breaking changes with some proposals. From my point of view, a breaking change is needed only when there are fatal design decisions which we want to drop or to allow completely new features which wouldn't have been possible otherwise. When it's possible to avoid them with a little trade-off, it's better because it eases adoption among existing devs and causes less friction.

Didn't had time to clone and experiment with your code yet, but I wanted to give you a feedback as soon as I could, so we could talk and make some decisions together.

Feel free to comment and integrate. This is meant as a PR review dialogue in written form. Everything can be openly discussed and changed 😉 I treat code as a way to achieve a goal, not as a personal creation, so don't worry to criticize and propose stuff.

## Breaking changes I noticed
- `SpeechRecognitionNotAvailableException`: if this gets reverted back to its original name `SpeechRecognitionNotAvailable`, it will not be a breaking change
- exceptions moved to `exception` package. It enhances code organization and it's good, but if the exceptions gets reverted to their original positions, this will not be a breaking change
- the static `init` method in `Speech` now requires a third parameter. By adding an additional init without the third parameter required, this will not be a breaking change

## Design decisions
- The constructors situation is as follows:
    ```java
    private Speech(final Context context)
    private Speech(final Context context, final String callingPackage)
    public Speech(final Context context, final String callingPackage, TextToSpeech.OnInitListener onInitListener)
   public Speech(final Context context, final String callingPackage, TextToSpeech.OnInitListener onInitListener, SpeechRecognitionListener speechRecognitionListener)
    public Speech(final Context context, final String callingPackage, TextToSpeech.OnInitListener onInitListener, SpeechRecognitionListener speechRecognitionListener, TextToSpeechListener textToSpeechListener)
    ```
    so we have 2 private constructors + 3 new public constructors + the static init which now has an additional required parameter (breaking change):
    ```java
    public static Speech init(final Context context, final String callingPackage, TextToSpeech.OnInitListener onInitListener)
    ```
    so at this state of the art, one could either use the singleton `Speech` or create many new instances using the newly added constructors.

    Although I'm no fan of Singletons at all, when I designed the library, I thought that since we have a single TTS engine and a single SpeechRecognizer on the system, I would honor that and act as a wrapper which could be called from whatever view I wanted, using a single instance. I would like to know more about your motivations.

    At this moment I can see two ways in which this could go:
    - we leave `Speech` as a singleton, adding the required static `init` overloads and leaving only this constructor, but making it private (this honors single instances and it's not a breaking change):
        ```java
        private Speech(final Context context, final String callingPackage, TextToSpeech.OnInitListener onInitListener, SpeechRecognitionListener speechRecognitionListener, TextToSpeechListener textToSpeechListener)
        ```
    - we remove the singleton and leave all the constructors public. This is a breaking change, which allows one to create multiple instances, too.

- `SpeechRecognitionListener` + `TextToSpeechListener`. Those are basically the interfaces extracted from `BaseSpeechRecognitionListener` and `BaseTextToSpeechListener` classes. Those two classes are not actually listeners. They contain the current TTS and TextToSpeech integration logic and so they need different names. I can see the potential in this. It allows customizations over the existing implementations and opens up for completely new implementations using also third party TTS and TextToSpeech implementations, making `Speech` a Facade.

Stopping at this point for now without going in further detail. We have some stuff to talk about 😉

yuripourre commented:
Thank you so much for the feedback.

Thinking about the breaking changes, they are not really necessary, I will revert them.


Regarding Singleton while changing it I had the (crazy?) idea that would be cool to have multiple voices, I don't know if it is possible though.

About the names, yes, they should change they are not listeners at all.

I will try to re read your comments more carefully and answer them properly this weekend.

gotev commented:
While waiting for your full comments, although you told me Kotlin is not your thing, I encourage you to take a look here: https://github.com/gotev/android-speech/tree/2.0/speech/src/main/java/net/gotev/speech/tts

> Regarding Singleton while changing it I had the (crazy?) idea that would be cool to have multiple voices, I don't know if it is possible though.

In `2.0` branch I've already come to a point where I achieved multiple voices at once, by creating multiple instances of `TextToSpeechEngine.kt`. It's an experiment though, so some tests are needed. It goes against my initial thoughts and assumptions for 1.0: single TTS, single Text To Speech. It's also a way to find a proof by contradiction or completely sink the initial 1.x idea. If the experiments in `2.0` works, ideas and library design behind `1.0` are proven wrong and new paths are open to be explored.

The main concept in `2.0` is to separate TTS from Speech To Text (both are called `engines`, so we have `TextToSpeechEngine` and `SpeechToTextEngine`) and be able to compose different `engines` at runtime to create rich dialogs (you can set a different voice to different instances of `TextToSpeechEngine`)  and allow the user to interact with his voice as well, without this having to be a mess to develop for whom uses the library. I used the Kotlin concept of [coroutines](https://kotlinlang.org/docs/reference/coroutines-overview.html) to express a function which takes some time to complete, but does not lock the thread and avoids complex callback-hell mechanisms, by returning the value asynchronously, but allowing you to still write imperative code. The concept is similar to that of C#'s async/await.

yuripourre commented:
Hi @gotev! Thank you for the feedback, I followed your comments:

- Rollback of logs and exceptions (naming and packages)
- Renamed the "listeners" to engine (based on your idea for 2.0).
- Made all constructors private and added a new init method

Regarding the constructors, I was so focused in do the refactoring that I didn't realize they were private.

I don't plan to change much more in this branch, since it currently meets my needs but I would really appreciate your thougths about what else could be changed so it could be merged.

PS: Thank you for mentioning coroutines, I will study it deeper.

gotev commented:
@yuripourre great! There are only some minor things left to do before the merge can be done:
- keep a single private constructor (the one with all the parameters) and inject default values in the static inits where needed
- update gradle plugin and gradle to latest recommended version by Android Studio

I can take care of them if you allow me to [make changes to this PR](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) and then I will proceed with the merge.

Since this PR introduces some enhancements and does not break anything, the version is going to be `1.5.0`.

I would like you to prepare the changelog, which I will then use in the release description + a mention to you and this PR.

Thank you for the contribution!

yuripourre commented:
@gotev those are awesome news!
I always leave this option enabled, feel free to change it. The code is not exactly mine, btw, since you are reviewing.
I already removed the private constructors :D

gotev commented:
@yuripourre awesome work

gotev commented:
`1.5.0` is now live 🎉

yuripourre commented:
@gotev awesome! Can't wait to do some experiments.

I didn't realize that there was an issue to change the voice.

gotev commented:
@yuripourre you can also update to `1.6.0` which has full Android 11 support now

yuripourre commented:
@gotev so cool that now is possible to change the voice!

gotev commented:
It was possible also before 1.5.0, but to get the available voices, some additional and not so pretty coding was required
